More in-depth round robin (#211)
jaybee84/ml-in-rd: content/06.model-complexity.md

All of these contribute to low signal to noise ratio in rare disease datasets.
Thus, applying ML to rare disease data without addressing the aforementioned shortcomings may lead to models that have low reproducibility or are hard to interpret.

Class imbalance in datasets can be addressed using decision tree-based ensemble learning methods (e.g., random forests)[@doi:10.1007/s11634-019-00354-x] (Figure[@fig:3]a).

> Random forests use resampling (with replacement) based techniques to form a consensus about the important predictive features identified by the decision trees [@https://doi.org/10.1023/A:1010933404324; @doi:10.1186/1472-6947-13-134]. [TODO: consider re-evaluating these references now that we're not explicitly talking about bagging?]

Author: Taking out the bagging TODO
```suggestion
Random forests use resampling (with replacement) based techniques to form a consensus about the important predictive features identified by the decision trees [@https://doi.org/10.1023/A:1010933404324; @doi:10.1186/1472-6947-13-134].
```

Additional approaches like combining random forests with resampling without replacement can generate confidence intervals for the model predictions by iteratively exposing the models to incomplete datasets, mimicking real world cases where most rare disease datasets are incomplete [@doi:10.3390/genes11020226].
Resampling approaches are most helpful in constructing confidence intervals for algorithms that generate the same outcome every time they are run (i.e., deterministic models).
For decision trees that choose features at random for selecting a path to the outcome (i.e., are non-deterministic), resampling approaches can be helpful in estimating the reproducibility of the model. 

> In situations where decision tree-based ensemble methods fail when they are applied to rare disease datasets, cascade learning has been a viable alternative [@pmid:30815073]. [TODO: first reference is generally about decision trees in imbalanced situations, not about cascade learning per se -- consider moving elsewhere in the sentence?]

Author: ```suggestion
In situations where decision tree-based ensemble methods fail when they are applied to rare disease datasets, cascade learning has been a viable alternative [@pmid:30815073].
```

Thus, applying ML to rare disease data without addressing the aforementioned shortcomings may lead to models that have low reproducibility or are hard to interpret.

> Class imbalance in datasets can be addressed using decision tree-based ensemble learning methods (e.g., random forests) (Figure[@fig:3]a).

Reviewer: ```suggestion
Class imbalance in datasets can be addressed using decision tree-based ensemble learning methods (e.g., random forests)[@doi:10.1007/s11634-019-00354-x] (Figure[@fig:3]a).
```

> Random forests use resampling (with replacement) based techniques to form a consensus about the important predictive features identified by the decision trees [@https://doi.org/10.1023/A:1010933404324; @doi:10.1186/1472-6947-13-134; @doi:10.1214/aos/1031689014; @doi:10.1177/2045894019890549; @doi:10/btzfh6]. [TODO: consider re-evaluating these references now that we're not explicitly talking about bagging?]

Author: A lot of these references are about bagging specifically. Do we want to take them out/reconsider now that this section has been modified?

Reviewer 2: Yes, good catch!

```suggestion
Random forests use resampling (with replacement) based techniques to form a consensus about the important predictive features identified by the decision trees [@https://doi.org/10.1023/A:1010933404324; @doi:10.1186/1472-6947-13-134; @doi:10.1214/aos/1031689014; @doi:10.1177/2045894019890549; @doi:10/btzfh6]. [TODO: consider re-evaluating these references now that we're not explicitly talking about bagging?]
```
```suggestion
Random forests use resampling (with replacement) based techniques to form a consensus about the important predictive features identified by the decision trees [@https://doi.org/10.1023/A:1010933404324; @doi:10.1186/1472-6947-13-134]. [TODO: consider re-evaluating these references now that we're not explicitly talking about bagging?]
```

Additional approaches like combining random forests with resampling without replacement can generate confidence intervals for the model predictions by iteratively exposing the models to incomplete datasets, mimicking real world cases where most rare disease datasets are incomplete [@doi:10.3390/genes11020226].
Resampling approaches are most helpful in constructing confidence intervals for algorithms that generate the same outcome every time they are run (i.e., deterministic models).
For decision trees that choose features at random for selecting a path to the outcome (i.e., are non-deterministic), resampling approaches can be helpful in estimating the reproducibility of the model. 

> In situations where decision tree-based ensemble methods fail when they are applied to rare disease datasets, cascade learning has been a viable alternative [@doi:10.1007/s11634-019-00354-x; @pmid:30815073]. [TODO: first reference is generally about decision trees in imbalanced situations, not about cascade learning per se -- consider moving elsewhere in the sentence?]

Author: Another question about a reference here.

Reviewer 2: I think we should remove it entirely or to another sentence, it seems irrelevant to the point of this sentence?

Reviewer: ```suggestion
In situations where decision tree-based ensemble methods fail when they are applied to rare disease datasets, cascade learning has been a viable alternative [@pmid:30815073]. [TODO: first reference is generally about decision trees in imbalanced situations, not about cascade learning per se -- consider moving elsewhere in the sentence?]
```

Reviewer: moved it to the section above where we talk about imbalanced datasets

In cascade learning, multiple methods leveraging distinct underlying assumptions are used in tandem to capture stable patterns existing in the dataset [@doi:10.1109/cvpr.2001.990537; @doi:10.1007/978-3-540-75175-5_16; @doi:10.1109/icpr.2004.1334680]. 
For example, a cascade learning approach for identifying rare disease patients from electronic health record data incorporated independent steps for feature extraction (word2vec [@arxiv:1301.3781]), preliminary prediction with ensembled decision trees, and then prediction refinement using data similarity metrics [@pmid:30815073]. 

> Combining these three methods resulted in better overall prediction when implemented on a silver standard dataset, as compared to a model that used ensemble-based prediction alone. [TODO: my understanding of the baseline model is that there was no feature selection or prediction refinement, but that prediction refinement did not get tested alone as the earlier language implied -- check my understanding]

Author: I made a change based on my read of the cited paper – do folks agree? cc: @jaybee84 in particular

Reviewer 2: Yes, I agree. I think I made this change in my last PR, but your wording is more accurate.

Reviewer: agree with the modifications here.

Author: Removing the TODO
```suggestion
Combining these three methods resulted in better overall prediction when implemented on a silver standard dataset, as compared to a model that used ensemble-based prediction alone.
```

> In addition to cascade learning, other approaches that can better represent rare classes like inverse class weighting [TODO: Add ref?] and oversampling [doi:10.1613/jair.953] may also help mitigate limitations due to class imbalance. 

Author: Should we add a reference for inverse class weighting specifically?

Reviewer 2: Yes, and also should we call this inverse probability weighting? Seems to be the more common term.

Author: Noting that I believe addressing these comments about (what we are currently calling) inverse class weighting is the last thing we wanted to do for _this pull request_. My understanding is that @jaybee84 will make suggestions here on this branch/PR.

Reviewer: ```suggestion
In addition to cascade learning, other approaches that can better represent rare classes using class re-balancing techniques like inverse sampling probability weighting [@doi:10.1186/s12911-021-01688-3], inverse class frequency weighting [@doi:10.1197/jamia.M3095], oversampling of rare classes [doi:10.1613/jair.953], or undersampling of majority class [@doi:10.48550/arXiv.1608.06048] may also help mitigate limitations due to class imbalance. 
```

The presence of label-noise and sparsity in the data can lead to overfitting of models to the training data, meaning that the models show high prediction accuracy on the training data but low prediction accuracy (and large prediction errors) on new evaluation data. 

> Overfit models tend to rely on patterns that are unique to the training data (for example, the calibration of the instrument that was used to generate the training data), and not generalizable to new data (e.g., data generated on the same instrument that has been recalibrated). [@doi:10.1073/pnas.1900654116] [TODO: ref is about interpretability, little mention of generalizability outside discussion of use of accuracy?]

Author: Question about reference.

Reviewer 2: ```suggestion
Overfit models tend to rely on patterns that are unique to the training data (for example, the clinical vocabulary or clinical coding practices at a specific hosptal), and not generalizable to new data (e.g., data collected at different hospitals). [@isbn:0262035618; @pmc:PMC8238368]
```

Multiple papers all cite this book for "overfitting". Not surprising, since it's such a general concept that there's probably no one good canonical paper to describe it. 

I also changed the example to something that is more relevant to disease.

Regularization can help in these scenarios.
Regularization is an approach by which a penalty or constraint is added to a model to avoid making large prediction errors.
These procedures can not only protect ML models as well as learned representations from poor generalizability caused by overfitting, but also reduce model complexity by reducing the feature space available for training [@doi:10.1371/journal.pgen.1004754, @doi:10.1002/sim.6782]. (Figure[@fig:3]a)

> Some examples of ML methods with regularization include ridge regression, LASSO regression, and elastic net regression [@doi:10.1111/j.1467-9868.2005.00503.x], among others.

Author: A little bit of a reorganization so we define regularization before we talk about what it can _specifically_ help with.

Reviewer 2: Looks great! 27 and 28 could possibly be combined ,but I don't think it's that important.

Regularization is often used in rare variant discovery and immune cell signature discovery studies; much like rare disease, these examples need to accommodate sparsity in data.
For example, LASSO has been used to capturing combinations of rare and common variants associated with specific traits have proven beneficial [@doi:10.1186/1753-6561-5-s9-s113].
In this example, applying LASSO regularization reduced the number of common variants included as features in the final analysis generating a simpler model while reducing error in the association of common and rare variants with a specific trait. 
In the context of rare immune cell signature discovery, variations of elastic-net regression were found to outperform other regression approaches [@doi:10.1016/j.compbiomed.2015.10.008; @doi:10.1186/s12859-019-2994-z]. 
Thus, regularization methods like LASSO or elastic-net are beneficial in ML with rare observations, and are worth exploring in the context of rare diseases.[@doi:10.1371/journal.pgen.1004754] 
Other examples of regularization that have been successfully applied to rare disease ML include using Kullback–Leibler (KL) divergence loss or dropout during neural network training.

> In a study using a variational autoencoder (VAE) (see Box 2: Definitions) for dimensionality reduction in gene expression data from acute myeloid leukemia (AML) samples, the KL loss between the input data and its low dimensional representation provided the regularizing penalty for the model. [@doi:10.1101/278739; @doi:10.48550/arXiv.1312.6114] [TODO: consider removing this sentence while retaining the mention of KL divergence loss in the previous sentence and only covering this in multiple approaches section?]

Author: Thoughts on this TODO?

Reviewer: this sentence is an example of where KL divergence loss was used in the context of rare disease. I think we can probably retain this brief mention here. I dont think we talk about it in the multiple approaches section much at all in our current version.

Author: ```suggestion
In a study using a variational autoencoder (VAE) (see Box 2: Definitions) for dimensionality reduction in gene expression data from acute myeloid leukemia (AML) samples, the KL loss between the input data and its low dimensional representation provided the regularizing penalty for the model. [@doi:10.1101/278739; @doi:10.48550/arXiv.1312.6114]
```

In a study using a convolutional neural network (CNN) to identify tubers in MRI images from TS patients, overfitting was minimized using the dropout regularization method which removed randomly chosen network nodes in each iteration of the CNN model generating simpler models in each iteration.[@doi:10.1371/journal.pone.0232376]
Thus, depending on the learning method that is used, regularization approaches should be incorporated into data analysis when working with rare disease datasets. 

> ![TODO: OLD FIGURE (new figure still WIP) Strategies to simplify models and stabilize predictions preserve the value of machine learning in rare disease. A-B) Strategies to build confidence in model predictions; A) A schematic showing the concept of bootstrap, B) A schematic showing the concept of ensemble learning to converge on reliable models; C-D) Strategies to simplify models by penalizing complexity in ML models; C) A schematic showing the concept of regularization to selectively learn relevant features, D) A schematic showing the concept of one-class-at-a-time learning to select few features at a time. Horizontal bars represent health of a model, models are represented as a network of nodes (features) and edges (relationships), nodes with solid edges represent real patterns, nodes with broken edges represent spurious patterns](images/figures/pdfs/statistical-techniques.png){#fig:3}

Author: Wanted to make sure to track this figure legend update.
